File: src/importer.rs
//! The `Importer` trait plus the built-in filesystem importer (`FsImporter`)
//! and its dart-faithful partial / index / import-only resolver.

use std::path::{Path, PathBuf};

use crate::Syntax;

/// An opaque, importer-defined canonical identifier for a resolved stylesheet.
///
/// Two URLs that canonicalize to the same `CanonicalUrl` are the SAME loaded
/// file — it is the module-cache / dedup key. What "canonical" *means* is the
/// importer's to define; sasso only uses it as an `Eq`/`Hash` key. For
/// [`FsImporter`] it is the absolute filesystem path. The field is private so
/// constraints/normalization can be added later without an API break.
#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub struct CanonicalUrl(String);

impl CanonicalUrl {
    /// Wrap an importer-chosen canonical identifier.
    pub fn new(url: impl Into<String>) -> Self {
        CanonicalUrl(url.into())
    }

    /// The underlying string.
    pub fn as_str(&self) -> &str {
        &self.0
    }
}

/// The source an [`Importer::load`] produced for a [`CanonicalUrl`] — dart-sass's
/// `ImporterResult`.
#[derive(Clone, Debug)]
pub struct ImporterResult {
    /// The stylesheet source text.
    pub contents: String,
    /// The syntax `contents` should be parsed with (SCSS / indented / plain CSS).
    pub syntax: Syntax,
    /// The URL to record for this source in generated source maps (dart-sass
    /// `ImporterResult.sourceMapUrl`); `None` falls back to the canonical URL.
    pub source_map_url: Option<String>,
}

/// A real importer failure — an I/O error, a permission error, an ambiguous
/// match, an invalid URL, … — as distinct from a plain miss. Surfaced as an
/// actionable compile [`crate::Error`], never silently treated as "not found".
#[derive(Clone, Debug)]
pub struct ImporterError {
    /// Human-readable description, used as the compile error message.
    pub message: String,
}

/// Context passed to [`Importer::canonicalize`] (dart-sass's `CanonicalizeContext`).
pub struct CanonicalizeContext<'a> {
    /// `true` when resolving an `@import` (which additionally considers
    /// import-only `*.import.{scss,sass}` files), `false` for `@use`/`@forward`.
    pub from_import: bool,
    /// The canonical URL of the stylesheet containing the rule being resolved,
    /// if any (`None` only when there is no containing file). Relative URLs
    /// resolve against it.
    pub containing_url: Option<&'a CanonicalUrl>,
}

/// Resolves `@use` / `@forward` / `@import` URLs in dart-sass's two phases.
///
/// [`canonicalize`](Importer::canonicalize) maps a (possibly relative,
/// extension-less) URL to a stable canonical identity WITHOUT loading it (the
/// result is the module-cache key); [`load`](Importer::load) then fetches that
/// identity's source. Both methods distinguish three outcomes: `Ok(Some(_))` =
/// handled; `Ok(None)` = this importer doesn't handle the URL; `Err(_)` =
/// handled but failed (a real, reportable error).
///
/// Implementing this gives the caller full control over where partials come
/// from — and, crucially, keeps all file access on the caller's side of a
/// sandbox boundary.
pub trait Importer {
    /// Map `url` to its canonical identity, or `Ok(None)` if this importer
    /// cannot resolve it. MUST NOT load the file.
    fn canonicalize(
        &self,
        url: &str,
        ctx: &CanonicalizeContext<'_>,
    ) -> Result<Option<CanonicalUrl>, ImporterError>;

    /// Load the source for a [`CanonicalUrl`] previously returned by
    /// [`canonicalize`](Importer::canonicalize); `Ok(None)` if it can no longer
    /// be found.
    fn load(&self, canonical: &CanonicalUrl) -> Result<Option<ImporterResult>, ImporterError>;
}

/// A filesystem [`Importer`] resolving Sass partials (`_name.scss`,
/// `name/_index.scss`) against a list of load paths.
///
/// Intended for CLI/standalone use; an embedder inside a sandbox should
/// supply its own [`Importer`] that routes through the sandbox's file
/// capability instead of touching the disk directly.
pub struct FsImporter {
    load_paths: Vec<PathBuf>,
}

impl FsImporter {
    /// Create an importer that searches `load_paths` in order.
    pub fn new(load_paths: Vec<PathBuf>) -> Self {
        FsImporter { load_paths }
    }
}

impl Importer for FsImporter {
    fn canonicalize(
        &self,
        url: &str,
        ctx: &CanonicalizeContext<'_>,
    ) -> Result<Option<CanonicalUrl>, ImporterError> {
        // dart-sass resolves a relative URL against the containing file's
        // directory first, then the configured load paths. A `CanonicalUrl`
        // here IS the absolute fs path, so the containing dir is its parent; a
        // containing URL with no parent (a bare entry name like `input.scss`)
        // means the current directory — reproducing the old base of
        // `dirname_of(url).unwrap_or_default()` (an empty string => CWD).
        //
        // NB: this `parent()`-based dirname must stay consistent with the
        // evaluator's `dirname_of` (eval/mod.rs), which derives
        // `current_file_dir` (the `@import` cache key) from the same canonical
        // URL — both use `Path::parent` + treat an empty parent as CWD.
        let base_dir: PathBuf = match ctx.containing_url {
            Some(c) => match Path::new(c.as_str()).parent() {
                Some(par) if !par.as_os_str().is_empty() => par.to_path_buf(),
                _ => PathBuf::new(),
            },
            None => PathBuf::new(),
        };
        let bases = std::iter::once(base_dir).chain(self.load_paths.iter().cloned());
        for base in bases {
            // `@use`/`@forward` (`from_import == false`) never consider
            // `.import` files (those are an `@import`-only escape hatch).
            match resolve_in_base(&base, url, ctx.from_import) {
                Resolution::Found(p) => {
                    // The canonical key is the resolved absolute path so the
                    // same file loaded via different URLs is cached once.
                    let key = std::fs::canonicalize(&p)
                        .map(|c| c.to_string_lossy().into_owned())
                        .unwrap_or_else(|_| p.to_string_lossy().into_owned());
                    return Ok(Some(CanonicalUrl::new(key)));
                }
                // An ambiguous match is an error in dart-sass; we preserve the
                // existing behavior of surfacing it as a miss (the eval layer
                // turns that into the import error) rather than an `Err`.
                Resolution::Ambiguous => return Ok(None),
                Resolution::NotFound => {}
            }
        }
        Ok(None)
    }

    fn load(&self, canonical: &CanonicalUrl) -> Result<Option<ImporterResult>, ImporterError> {
        let p = Path::new(canonical.as_str());
        match std::fs::read_to_string(p) {
            Ok(contents) => Ok(Some(ImporterResult {
                contents,
                syntax: syntax_for_path(p),
                source_map_url: None,
            })),
            // The file vanished between `canonicalize` and `load` -> a miss.
            Err(e) if e.kind() == std::io::ErrorKind::NotFound => Ok(None),
            // Any other read failure (permission, invalid UTF-8, I/O) is a real,
            // reportable error via the `ImporterError` channel, not a misleading
            // "can't find stylesheet".
            Err(e) => Err(ImporterError {
                message: format!("Cannot read {}: {e}", p.display()),
            }),
        }
    }
}

/// The syntax a resolved file should be parsed with, from its extension: a
/// `.sass` file is the indented syntax, anything else (`.scss`) is SCSS.
fn syntax_for_path(p: &Path) -> Syntax {
    match p.extension().and_then(|e| e.to_str()) {
        Some(e) if e.eq_ignore_ascii_case("sass") => Syntax::Sass,
        Some(e) if e.eq_ignore_ascii_case("css") => Syntax::Css,
        _ => Syntax::Scss,
    }
}

/// Outcome of resolving an `@import` argument within a single load path.
enum Resolution {
    /// Exactly one candidate file matched.
    Found(PathBuf),
    /// Two or more candidates matched at the same precedence tier — dart-sass
    /// treats this as an error ("It's not clear which file to import.").
    Ambiguous,
    /// No candidate matched in this base directory.
    NotFound,
}

/// Resolve `@import "path"` against `base`, following dart-sass precedence.
///
/// dart-sass tries, in strict order (each "tier" is checked together; if a
/// tier has more than one match it is ambiguous, if it has exactly one that
/// wins, otherwise fall through to the next tier):
///
/// 1. import-only, non-index: `name.import.{scss,sass}` + partials
/// 2. normal, non-index:      `name.{scss,sass}` + partials
/// 3. import-only index:      `name/index.import.{…}` + `_index.import.{…}`
/// 4. normal index:           `name/index.{…}` + `_index.{…}`
///
/// An explicit `.scss`/`.sass` extension keeps only the matching extension but
/// still honours the import-only override (`name.import.scss`).
///
/// We deliberately do **not** resolve plain `.css` files here: importing a CSS
/// file as a stylesheet is a distinct feature (with its own strict-CSS parsing
/// rules) that this build doesn't implement, and treating `foo.css` as a
/// stylesheet would mis-handle constructs that dart-sass rejects.
///
/// An import-only file whose body is only `@forward`/`@use` (the real-world
/// shape — re-exporting another module) is treated as unusable and skipped,
/// since this build has no module system; resolution then falls through to the
/// normal file, matching the output we can actually produce.
fn resolve_in_base(base: &Path, path: &str, allow_import_only: bool) -> Resolution {
    // dart-sass normalizes the URL lexically before touching the filesystem,
    // so `foo/bar/../baz` resolves even when `foo/bar` doesn't exist.
    let normalized = lexical_normalize(path);
    let path = normalized.as_str();
    let p = Path::new(path);
    let dir = match p.parent() {
        Some(par) if !par.as_os_str().is_empty() => base.join(par),
        _ => base.to_path_buf(),
    };
    let file = p.file_name().and_then(|s| s.to_str()).unwrap_or(path);

    // Explicit `.css` extension: only the plain-CSS candidate is considered.
    // (An `@import "x.css"` never reaches here — it's a passthrough upstream —
    // but `@use "x.css"` does.)
    if let Some(stem) = file.strip_suffix(".css") {
        return match tier_exact(&dir, stem, &["css"], false) {
            Tier::One(p) => Resolution::Found(p),
            Tier::Many => Resolution::Ambiguous,
            Tier::None => Resolution::NotFound,
        };
    }

    // Explicit `.scss`/`.sass` extension: only that extension is considered.
    let explicit_ext = [".scss", ".sass"].into_iter().find(|ext| file.ends_with(ext));

    if let Some(ext) = explicit_ext {
        let stem = &file[..file.len() - ext.len()];
        // Strip the leading dot so `ext` is e.g. "scss".
        let ext = &ext[1..];
        // Tier 1: import-only override for the explicit extension.
        if allow_import_only {
            match tier_exact(&dir, stem, &[ext], true) {
                Tier::One(p) => return Resolution::Found(p),
                Tier::Many => return Resolution::Ambiguous,
                Tier::None => {}
            }
        }
        // Tier 2: the file as written (+ partial).
        return match tier_exact(&dir, stem, &[ext], false) {
            Tier::One(p) => Resolution::Found(p),
            Tier::Many => Resolution::Ambiguous,
            Tier::None => Resolution::NotFound,
        };
    }

    // Extensionless: scss/sass are equal precedence, css is a fallback.
    let mut non_index: Vec<(&str, bool)> = Vec::with_capacity(2);
    if allow_import_only {
        non_index.push((file, true));
    }
    non_index.push((file, false));
    for (stem, import_only) in &non_index {
        match tier_with_extensions(&dir, stem, *import_only) {
            Tier::One(p) => return Resolution::Found(p),
            Tier::Many => return Resolution::Ambiguous,
            Tier::None => {}
        }
    }

    // A plain `.css` file (loaded in plain-CSS mode): after the Sass
    // candidates, but BEFORE index files (dart-sass `_tryPathWithExtensions`
    // tries `$path.css` before `_tryPathAsDirectory`).
    match tier_exact(&dir, file, &["css"], false) {
        Tier::One(p) => return Resolution::Found(p),
        Tier::Many => return Resolution::Ambiguous,
        Tier::None => {}
    }

    // Index files live in a subdirectory named after the import path.
    let index_dir = dir.join(file);
    let index_modes: &[bool] = if allow_import_only {
        &[true, false]
    } else {
        &[false]
    };
    for import_only in index_modes {
        match tier_with_extensions(&index_dir, "index", *import_only) {
            Tier::One(p) => return Resolution::Found(p),
            Tier::Many => return Resolution::Ambiguous,
            Tier::None => {}
        }
    }

    Resolution::NotFound
}

/// Lexically remove `.` and `..` segments from a URL path (no filesystem
/// access; leading `..` segments that would escape are kept).
fn lexical_normalize(path: &str) -> String {
    let mut out: Vec<&str> = Vec::new();
    for seg in path.split('/') {
        match seg {
            "" | "." => {}
            ".." => {
                if matches!(out.last(), Some(&s) if s != "..") {
                    out.pop();
                } else {
                    out.push("..");
                }
            }
            s => out.push(s),
        }
    }
    let mut s = out.join("/");
    if path.starts_with('/') {
        s.insert(0, '/');
    }
    if s.is_empty() {
        s.push('.');
    }
    s
}

/// One precedence tier's matches.
enum Tier {
    None,
    One(PathBuf),
    Many,
}

impl Tier {
    fn from(mut found: Vec<PathBuf>) -> Tier {
        match found.len() {
            0 => Tier::None,
            1 => Tier::One(found.pop().unwrap_or_default()),
            _ => Tier::Many,
        }
    }
}

/// Try a tier with the standard extension grouping: `scss` and `sass` are
/// checked together (any match there wins the tier). Each extension is checked
/// in both non-partial and partial (`_name`) forms.
fn tier_with_extensions(dir: &Path, stem: &str, import_only: bool) -> Tier {
    tier_exact(dir, stem, &["scss", "sass"], import_only)
}

/// Collect existing candidate files for `stem` under `dir` across `exts`, in
/// non-partial and partial forms, optionally inserting the `.import` suffix
/// (import-only files). Returns how many matched.
fn tier_exact(dir: &Path, stem: &str, exts: &[&str], import_only: bool) -> Tier {
    let mut found = Vec::new();
    let suffix = if import_only { ".import" } else { "" };
    for ext in exts {
        for name in [format!("_{stem}{suffix}.{ext}"), format!("{stem}{suffix}.{ext}")] {
            let cand = dir.join(&name);
            if cand.is_file() {
                found.push(cand);
            }
        }
    }
    Tier::from(found)
}

File: wasm/npm/_importer.mjs
// Importer machinery for the `sasso` npm package.
//
// wasm has no filesystem, so every `@use`/`@forward`/`@import` is resolved here
// in JS and bridged into the engine through the wasm host functions (see
// `_loader.mjs`). This module provides the two resolver kinds sasso merges into
// one chain per compile:
//
//   • a Node-fs importer (`makeFsImporter`) for `loadPaths` and relative-to-
//     containing-file resolution — a faithful JS port of the dart-sass
//     partial / index / import-only precedence in `../../src/importer.rs`; and
//   • a bridge (`normalizeImporter`) for user-supplied dart-sass *modern*
//     importers — both `{ canonicalize, load }` Importers and `{ findFileUrl }`
//     FileImporters.
//
// The internal resolver interface is sync and string-based:
//   canonicalize(url, fromImport, containingHref|null) -> canonicalHref|null
//   load(canonicalHref) -> { contents, syntax: 0|1|2, sourceMapUrl: string|null } | null
// `syntax`: 0 = SCSS, 1 = indented `.sass`, 2 = plain CSS.
//
// **Synchronous only.** The engine is sync, so importer callbacks cannot await;
// a user importer that returns a Promise throws a clear error (even under
// `compileStringAsync`). Truly-async importers are unsupported.

import { existsSync, statSync, readFileSync, realpathSync } from "node:fs";
import { pathToFileURL, fileURLToPath } from "node:url";
import * as nodePath from "node:path";

const SYNTAX_SCSS = 0;
const SYNTAX_SASS = 1;
const SYNTAX_CSS = 2;

function isThenable(x) {
  return x != null && typeof x.then === "function";
}

const ASYNC_UNSUPPORTED =
  "sasso: asynchronous importers are not supported — the wasm engine is " +
  "synchronous, so importer callbacks must return synchronously (even under " +
  "compileStringAsync).";

/** Map a dart-sass syntax string to the wasm syntax code. */
export function syntaxCode(syntax) {
  if (syntax === "indented" || syntax === "sass") return SYNTAX_SASS;
  if (syntax === "css") return SYNTAX_CSS;
  return SYNTAX_SCSS;
}

/** The syntax code for a resolved file path, from its extension. */
export function syntaxForPath(p) {
  const ext = nodePath.extname(p).toLowerCase();
  if (ext === ".sass") return SYNTAX_SASS;
  if (ext === ".css") return SYNTAX_CSS;
  return SYNTAX_SCSS;
}

function isFile(p) {
  try {
    return statSync(p).isFile();
  } catch {
    return false;
  }
}

// --- dart-sass filesystem resolution (port of src/importer.rs) -------------

/** Lexically remove `.` / `..` segments from a URL path (no fs access). */
function lexicalNormalize(path) {
  const out = [];
  for (const seg of path.split("/")) {
    if (seg === "" || seg === ".") continue;
    if (seg === "..") {
      if (out.length && out[out.length - 1] !== "..") out.pop();
      else out.push("..");
    } else {
      out.push(seg);
    }
  }
  let s = out.join("/");
  if (path.startsWith("/")) s = "/" + s;
  if (s === "") s = ".";
  return s;
}

// One precedence tier: collect existing candidates, returning a single match,
// `"ambiguous"` for >1 at the same tier, or `null` for none.
function tierExact(dir, stem, exts, importOnly) {
  const found = [];
  const suffix = importOnly ? ".import" : "";
  for (const ext of exts) {
    for (const name of [`_${stem}${suffix}.${ext}`, `${stem}${suffix}.${ext}`]) {
      const cand = nodePath.join(dir, name);
      if (isFile(cand)) found.push(cand);
    }
  }
  if (found.length === 0) return null;
  if (found.length > 1) return "ambiguous";
  return found[0];
}

function tierWithExtensions(dir, stem, importOnly) {
  return tierExact(dir, stem, ["scss", "sass"], importOnly);
}

// Resolve `path` against `base` following dart-sass precedence. Returns an
// absolute path, the sentinel `"ambiguous"`, or `null` (not found here).
function resolveInBase(base, path, allowImportOnly) {
  const normalized = lexicalNormalize(path);
  const parsed = nodePath.posix.parse(normalized.replace(/\\/g, "/"));
  // The directory portion of the (normalized) import path, joined onto `base`.
  const subDir = parsed.dir && parsed.dir !== "" ? parsed.dir : "";
  const dir = subDir ? nodePath.join(base, subDir) : base;
  const file = parsed.base || normalized;

  // Explicit `.css`: only the plain-CSS candidate.
  if (file.endsWith(".css")) {
    return tierExact(dir, file.slice(0, -4), ["css"], false);
  }

  // Explicit `.scss`/`.sass`: only that extension (+ import-only override).
  const explicitExt = [".scss", ".sass"].find((e) => file.endsWith(e));
  if (explicitExt) {
    const stem = file.slice(0, -explicitExt.length);
    const ext = explicitExt.slice(1);
    if (allowImportOnly) {
      const r = tierExact(dir, stem, [ext], true);
      if (r) return r;
    }
    return tierExact(dir, stem, [ext], false);
  }

  // Extensionless: scss/sass equal precedence, then css, then index dirs.
  const nonIndex = [];
  if (allowImportOnly) nonIndex.push([file, true]);
  nonIndex.push([file, false]);
  for (const [stem, importOnly] of nonIndex) {
    const r = tierWithExtensions(dir, stem, importOnly);
    if (r) return r;
  }

  const cssr = tierExact(dir, file, ["css"], false);
  if (cssr) return cssr;

  const indexDir = nodePath.join(dir, file);
  const indexModes = allowImportOnly ? [true, false] : [false];
  for (const importOnly of indexModes) {
    const r = tierWithExtensions(indexDir, "index", importOnly);
    if (r) return r;
  }

  return null;
}

/** Canonical key for a resolved path: its realpath as a `file:` URL href. */
function canonicalHrefFor(path) {
  let real = path;
  try {
    real = realpathSync(path);
  } catch {
    // keep the un-canonicalized path (mirrors src/importer.rs's fallback)
  }
  return pathToFileURL(real).href;
}

/** Read a resolved file as an importer result (`null` if it vanished). */
function loadFsPath(path) {
  let contents;
  try {
    contents = readFileSync(path, "utf8");
  } catch (e) {
    if (e && e.code === "ENOENT") return null; // raced between resolve and load
    throw new Error(`Cannot read ${path}: ${e && e.message ? e.message : e}`);
  }
  return { contents, syntax: syntaxForPath(path), sourceMapUrl: null };
}

/**
 * A Node-fs importer searching, in order, the containing file's directory then
 * `loadPaths`, with dart-faithful partial/index/import-only precedence.
 */
export function makeFsImporter(loadPaths) {
  const bases = (loadPaths || []).map((p) => String(p));
  return {
    canonicalize(url, fromImport, containingHref) {
      // Base directories: the containing file's dir (when it is a file: URL),
      // then the configured load paths. Unlike the CLI's FsImporter we do NOT
      // fall back to the CWD when there is no containing file — dart-sass's
      // `compileString` only resolves relative URLs when given a `url` (or via
      // `loadPaths`), so an import with neither simply misses.
      const baseDirs = [];
      if (containingHref) {
        try {
          baseDirs.push(nodePath.dirname(fileURLToPath(containingHref)));
        } catch {
          // containing URL isn't a file: URL — skip relative resolution
        }
      }
      for (const b of bases) baseDirs.push(b);

      for (const base of baseDirs) {
        const r = resolveInBase(base, url, fromImport);
        if (r === "ambiguous") return null; // dart errors; we treat as a miss
        if (r) return canonicalHrefFor(r);
      }
      return null;
    },
    load(canonicalHref) {
      let path;
      try {
        path = fileURLToPath(canonicalHref);
      } catch {
        return null;
      }
      return loadFsPath(path);
    },
  };
}

// --- user (dart-sass modern) importer bridging -----------------------------

function ctxFor(fromImport, containingHref) {
  return {
    fromImport,
    containingUrl: containingHref ? new URL(containingHref) : undefined,
  };
}

function toHref(urlOrString) {
  return urlOrString instanceof URL ? urlOrString.href : new URL(urlOrString).href;
}

// Settle a possibly-thenable user return into the chain interface. In `async`
// mode the result is always a Promise (so async importers are awaited by the
// async engine); in sync mode a Promise is a hard error (the sync engine can't
// await it), and a plain value maps through immediately.
function settle(raw, map, async) {
  if (async) return Promise.resolve(raw).then((v) => (v == null ? null : map(v)));
  if (isThenable(raw)) throw new Error(ASYNC_UNSUPPORTED);
  return raw == null ? null : map(raw);
}

const loadMap = (r) => ({
  contents: r.contents,
  syntax: syntaxCode(r.syntax),
  sourceMapUrl: r.sourceMapUrl != null ? String(r.sourceMapUrl) : null,
});

/** Wrap a user `{ canonicalize, load }` dart-sass Importer (`async` = await Promises). */
function wrapImporter(imp, async) {
  return {
    canonicalize(url, fromImport, containingHref) {
      return settle(imp.canonicalize(url, ctxFor(fromImport, containingHref)), toHref, async);
    },
    load(canonicalHref) {
      return settle(imp.load(new URL(canonicalHref)), loadMap, async);
    },
  };
}

/**
 * Wrap a user `{ findFileUrl }` dart-sass FileImporter: `findFileUrl` returns a
 * `file:` URL, which we then resolve on disk with the standard partial/index
 * precedence and read. (`async` = await an async `findFileUrl`.)
 */
function wrapFileImporter(imp, async) {
  const finish = (r, fromImport) => {
    if (r == null) return null;
    const fileUrl = r instanceof URL ? r : new URL(r);
    if (fileUrl.protocol !== "file:") {
      throw new Error(
        `sasso: FileImporter.findFileUrl must return a file: URL, got ${fileUrl.protocol}`,
      );
    }
    const target = fileURLToPath(fileUrl);
    const resolved = resolveInBase(nodePath.dirname(target), nodePath.basename(target), fromImport);
    return !resolved || resolved === "ambiguous" ? null : canonicalHrefFor(resolved);
  };
  return {
    canonicalize(url, fromImport, containingHref) {
      const raw = imp.findFileUrl(url, ctxFor(fromImport, containingHref));
      if (async) return Promise.resolve(raw).then((r) => finish(r, fromImport));
      if (isThenable(raw)) throw new Error(ASYNC_UNSUPPORTED);
      return finish(raw, fromImport);
    },
    load(canonicalHref) {
      try {
        return loadFsPath(fileURLToPath(canonicalHref));
      } catch {
        return null;
      }
    },
  };
}

/**
 * Normalize one user importer (Importer or FileImporter) to the chain interface.
 * Pass `async = true` for the asyncified engine (callbacks may return Promises);
 * the default (sync engine) rejects Promises with a clear error.
 */
export function normalizeImporter(imp, async = false) {
  if (imp && typeof imp.canonicalize === "function" && typeof imp.load === "function") {
    return wrapImporter(imp, async);
  }
  if (imp && typeof imp.findFileUrl === "function") {
    return wrapFileImporter(imp, async);
  }
  throw new Error(
    "sasso: each importer must be a dart-sass Importer ({ canonicalize, load }) " +
      "or FileImporter ({ findFileUrl }).",
  );
}
